pkgs/development/interpreters/python/mk-python-derivation.nix

```diff
@@ -104,6 +104,45 @@ else
 let
   inherit (python) stdenv;
 
+  validatePythonMatches = attrName: let
+    isPythonModule = drv:
+      # all pythonModules have the pythonModule attribute
+      (drv ? "pythonModule")
+      # Some pythonModules are turned in to a pythonApplication by setting the field to false
+      && (!builtins.isBool drv.pythonModule);
+    isMismatchedPython = drv: drv.pythonModule != python;
+
+    optionalLocation = let
+        pos = builtins.unsafeGetAttrPos (if attrs ? "pname" then "pname" else "name") attrs;
+      in if pos == null then "" else " in ${pos.file}:${toString pos.line}:${toString pos.column}";
+
+    leftPadName = name: against: let
+        len = lib.max (lib.stringLength name) (lib.stringLength against);
+      in lib.strings.fixedWidthString len " " name;
+
+    drvName = drv: drv.pname or drv.name;
+    warnMismatch = drv: lib.warn ''
+      Warning: Python version mismatch in '${name}':
+
+      The Python derivation '${name}' depends on a Python derivation
+      named '${drvName drv}', but the two derivations use different versions
+      of Python:
+
+          ${leftPadName name (drvName drv)} uses ${python}
+          ${leftPadName (drvName drv) name} uses ${toString drv.pythonModule}
+
+      Suggestion: change ${name}'s ${attrName} to use a '${drvName drv}'
+      built from the same version of Python${optionalLocation}.
+    ''
```

Comment:
Another good suggestion would be:

```
move the package from ${attrName} to nativeBuildInputs
```

in the case of, example, using python3's black on an python2 project.

Reply:
Maybe have a list of possible solutions?

```
Possible solutions:
  * change ${name}'s ${attrName} to use a '${drvName drv}'
    built from the same version of Python${optionalLocation}.

  * move ${drvName drv} from ${attrName} to nativeBuildInputs.
```

Reply:
Unless you have a way to determine exactly which suggestion should be shown :P